packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+): Promise<TTransaction & IFullySignedTransaction> {
+    const signedTransaction = await partiallySignTransactionWithSigners(transaction, signers);
+    assertTransactionIsFullySigned(signedTransaction);
+    return signedTransaction;
+}
+
+/**
+ * Signs and sends a transaction using any signers that may be stored in IAccountSignerMeta
+ * instruction accounts as well as any signers provided explicitly to this function.
+ * It will identify a single TransactionSendingSigners to use for sending the transaction, if any.
+ * Otherwise, it will send the transaction using the provided fallbackSender.
+ */
+export async function signAndSendTransactionWithSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    signers: readonly TransactionSigner[] = [],
+    fallbackSender?: (transaction: TTransaction & IFullySignedTransaction) => Promise<SignatureBytes>
+): Promise<SignatureBytes> {
+    const { partialSigners, modifyingSigners, sendingSigner } = categorizeTransactionSigners(
+        deduplicateSigners([...getSignersFromTransaction(transaction).filter(isTransactionSigner), ...signers])
+    );
+
+    const signedTransaction = await signModifyingAndPartialTransactionSigners(
+        transaction,
+        modifyingSigners,
+        partialSigners
+    );
+
+    if (sendingSigner) {
+        const [signature] = await sendingSigner.signAndSendTransactions([signedTransaction]);
+        return signature;
+    }
+
+    if (fallbackSender) {
+        assertTransactionIsFullySigned(signedTransaction);
+        return fallbackSender(signedTransaction);
+    }
+
+    // TODO: Coded error.
+    throw new Error('No TransactionSendingSigner was identified and no fallback sender was provided.');
+}
+
+/**
+ * Identifies each provided TransactionSigner and categorizes them into their respective types.
+ * When a signer implements multiple interface, it will try to used to most powerful interface
+ * but fallback to the least powerful interface when necessary.
+ * For instance, if a signer implements TransactionSendingSigner and TransactionModifyingSigner,
+ * it will be categorized as a TransactionSendingSigner if and only if no other signers implement
+ * the TransactionSendingSigner interface.
+ */
+function categorizeTransactionSigners(
+    signers: readonly TransactionSigner[],
+    identifySendingSigner: boolean = true
+): Readonly<{
+    modifyingSigners: readonly TransactionModifyingSigner[];
+    partialSigners: readonly TransactionPartialSigner[];
+    sendingSigner: TransactionSendingSigner | null;
+}> {
+    // Identify the unique sending signer that should be used.
+    const sendingSigner = identifySendingSigner ? identifyTransactionSendingSigner(signers) : null;
+
+    // Now, focus on the other signers.
+    // I.e. the modifying or partial signers that are not the identified sending signer.
+    // Note that any other sending only signers will be discarded.
```

Comment:
Shouldn't we fatal if there are more than one sending signer? That would be extremely surprising to me for one of my senders to get chosen, for reasons that are inscrutable to me, and not the other(s).

Reply:
We could fail but that will limit some of the advanced use-cases.

Again, the ephemeral signer example illustrates this nicely. Take [the example on this thread](https://github.com/solana-labs/solana-web3.js/pull/1792#discussion_r1394075990) and focus on the second bullet point, i.e. the ephemeral signers are implements such that they defer to the original signer. Now imagine that the original signer is a `TransactionSendingSigner`. All of the sudden you end up with multiple `TransactionSendingSigner` for different addresses (the ephemeral accounts) yet it does not matter which one is selected as they all execute the original signer's logic.

Reply:
![mild-headache-headache.gif](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lE1zD7Jpg6mWNv3djNKZ/dae4de4c-2a7d-4c9a-9eca-775b3b430a74.gif)

This will definitely get misused.

I also really don't like not being able to look at a transaction and predict which signer will send the transaction until you try.

Reply:
I'm happy for us to fail here but then we definitely cannot remove the optional `signers` array because we'd be making ephemeral signers impossible with the Signer API.

There are basically two ways the Signer API can currently support ephemeral signers depending on how the ephemeral signers themselves are implemented:

- **Option A: Delegate to the original signer.**
    
    One way to implement these ephemeral signers is to simply delegate to the `signAndSendTransactions` method of the original signer. The issue here is that we have 2 ephemeral signers. That being said, **if our sign and send helper doesn’t fail** and simply chooses one of the available `TransactionSendingSigner`, then that’ll work out of the box.
    
    ```tsx
    await signAndSendTransactionWithSigners(myTransaction);
    ```
    
- **Option B: Use noop signers.**
    
    Another way would be to make these ephemeral signers noop signers. Since we still need to use the original signer and it’s not part of the transaction, we would need to accept an optional array of additional signers to use in our sign and send helper function.
    
    ```tsx
    await signAndSendTransactionWithSigners(myTransaction, [myOriginalSigner]);
    ```

Reply:
Resolving this as a decision was made [here](https://github.com/solana-labs/solana-web3.js/pull/1792#issuecomment-1820487421).